### PR TITLE
Add All Metadata to Manifest

### PR DESCRIPTION
**Pull request by GPelayo**

### CODE REVIEW

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 1]
```diff
@@ -305,15 +321,20 @@ def _create_zipped_bdbag(participant_file_object: StringIO, sample_file_object:
             return bdbag_api.archive_bag(bag_path, 'zip')
 
     def _iterate_hit_tsv(self, es_search_hit, writer):
+        # if entity_type == 'files':
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: Remove

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 2]
```diff
@@ -305,15 +321,20 @@ def _create_zipped_bdbag(participant_file_object: StringIO, sample_file_object:
...
+        #     assert len(hit_dict['contents'][entity_type]) == 1
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: Remove

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 3]
```diff
@@ -305,15 +321,20 @@ def _create_zipped_bdbag(participant_file_object: StringIO, sample_file_object:
...
-        for bundle in hit_dict['bundles']:
-            # FIXME: If a file is in multiple bundles, the manifest will list it twice. `hca dss download_manifest`
-            # would download the file twice (https://github.com/DataBiosphere/azul/issues/423).
-            writer.writerow(self._translate(bundle, 'bundles') + file_fields)
+        inner_entity_fields = []
+        for doc_path, column_mapping in self.manifest_entries.items():
+            doc_path = doc_path.split('.')
+            entities = hit_dict
+            for key in doc_path:
+                entities = entities.get(key, [{}])
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: This works if the missing key is the last component of doc path. If earlier components were missing, like `contents` for example, the next loop iteration would attempt to do `[{}].get(key)` and raise an exception.

File: test/service/test_request_validation.py [hunk 4]
```diff
@@ -165,23 +165,43 @@ def test_file_order(self):
...
+                        url = self.base_url + '/repository/files/export?filters={"file":{}}'
+                        response = requests.get(url)
+                        self.assertEqual(200, response.status_code, 'Unable to download manifest')
+                        tsv_file = csv.DictReader(response.iter_lines(decode_unicode=True), delimiter='\t')
+                        rows = {r['file_uuid']: r for r in tsv_file}
+                        # 7 because self.bundle and bundle f79257a7-dfc6-46d6-ae00-ba4b25313c10 have 7 files total.
+                        self.assertEqual(len(rows.items()), 7, 'Wrong number of files were found.')
+                        # Fieldnames from the bundles entity are always the first headers in the manifest. The rest of
+                        # the fieldnames are ordered as they are mapped in request_config.json from top to bottom.
+                        expected_fieldnames = ["bundle_uuid", "bundle_version", "file_content_type", "file_name",
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: Checkout the bdbag test and how it asserts the first column literally. You could also assert the entire tsv literall in one giant assertEquals(). freeze() is your friend.

Comment (GPelayo) [reply test/service/test_request_validation.py]: I tried copying that test last week, but the test manifest generated isn't the same between tests. If a manifest with the exact same files is generated, its cells containing concatenated values (such as '123 || 345') are not written the same order. Should I sort these values before concatenating them?

Comment (hannes-ucsc) [reply test/service/test_request_validation.py]: I think that would be a good idea.

File: test/service/test_request_validation.py [hunk 5]
```diff
@@ -167,23 +167,101 @@ def test_file_order(self):
...
-            self.assertEqual(expected_fieldnames, tsv_file.fieldnames, 'Manifest headers are not configured correctly')
+            for single_part in "0", "1":
+                with self.subTest(is_single_part=single_part):
+                    with mock.patch.dict(os.environ, AZUL_DISABLE_MULTIPART_MANIFESTS=single_part):
+                        url = self.base_url + '/repository/files/export?filters={"file":{}}'
+                        response = requests.get(url)
+                        self.assertEqual(200, response.status_code, 'Unable to download manifest')
+
+                        expected_pdf = [
+                            ('bundle_uuid', 'f79257a7-dfc6-46d6-ae00-ba4b25313c10'),
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: If you really want to assert a second row, I would make these tuples triples and eliminate `expected_fastq`:

```suggestion
                            ('bundle_uuid', 'f79257a7-dfc6-46d6-ae00-ba4b25313c10', 'f79257a7-dfc6-46d6-ae00-ba4b25313c10'),                      
```

Some lines might require some creative wrapping but it's still going to look better than duplicating the headers in `expected_fastq` only to discard them later.

File: src/azul/service/service_config/request_config.json [hunk 6]
```diff
@@ -62,11 +59,35 @@
...
-      "bundle_uuid": "uuid",
-      "bundle_version": "version"
+    "contents.cell_suspensions": {
+      "total_estimated_cells": "total_estimated_cells"
+    },
+    "contents.protocols": {
+      "instrument_manufacturer_model": "instrument_manufacturer_model",
+      "library_construction_approach": "library_construction_approach"
+    },
+    "contents.projects": {
```
Comment (hannes-ucsc) [inline src/azul/service/service_config/request_config.json]: You missed a bunch of project fields like `project_short_name`.

File: src/azul/service/service_config/request_config.json [hunk 7]
```diff
@@ -62,11 +59,39 @@
...
+      "contact_names": "contact_names",
+      "document_id": "document_id",
+      "institution": "institution",
```
Comment (hannes-ucsc) [inline src/azul/service/service_config/request_config.json]:
```suggestion
      "institutions": "institutions",
```

File: test/service/test_request_validation.py [hunk 8]
```diff
@@ -168,23 +168,78 @@ def test_file_order(self):
...
-            tsv_file = csv.DictReader(response.iter_lines(decode_unicode=True), delimiter='\t')
-            # 2 because self.bundle has 2 files
-            self.assertEqual(len(list(tsv_file)), 2, 'Wrong number of files were found.')
-            manifest_config = json.load(open('{}/request_config.json'.format(self.service_config_dir), 'r'))['manifest']
-            expected_fieldnames = list(manifest_config['bundles'].keys()) + list(
-                manifest_config['contents.files'].keys())
-            self.assertEqual(expected_fieldnames, tsv_file.fieldnames, 'Manifest headers are not configured correctly')
+            for single_part in "0", "1":
+                with self.subTest(is_single_part=single_part):
+                    with mock.patch.dict(os.environ, AZUL_DISABLE_MULTIPART_MANIFESTS=single_part):
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: Can you try patching the corresponding Config property instead? The point of Config is not to sprinkle the environment variable names all over the code base.

Comment (hannes-ucsc) [reply test/service/test_request_validation.py]: You'll probably want to use `mock.patch.object(azul.config, …)`

**Comment by hannes-ucsc** [comment]
IT passes: https://gitlab.dev.explore.data.humancellatlas.org/azul/azul/-/jobs/316

**Comment by hannes-ucsc** [comment]
Feel free to push again, @GPelayo.

**Comment by hannes-ucsc** [comment]
One last push to Gitlab.

**Pull request merged**